### ISSUE TITLE
 [jvm-packages]Does xgboost on spark support the leanring to rank?

### ISSUE THREAD

**Issue by anddelu** [dmlc/xgboost#2654]
I didn't see the demo of the learning to rank in xgboost on spark environment. If it supports, how to specify the trainning data and group data.

**Comment by superbobry** [comment]
It doesn't at the moment. We would need to work on the API to pass group data, because the usual XGBoost way doesn't play nicely with RDD/DataFrame.

**Comment by superbobry** [comment]
I don't think this counts as support, it's fragile even for unit tests.

**Comment by CodingCat** [comment]
where `partition` is involved?

**Comment by CodingCat** [comment]
if you were talking about that TaskContext.PartitionId(), I think it's not to force an RDD's partition but how to construct that Seq[Seq[Int]], if you think in that way, it is much easier to move forward 

The biggest concern with the current approach is the scalability, I would like it to be the similar approach with baseMargin/weight support

**Comment by anddelu** [comment]
thanks all. I am still not clear about some quetions. 
1: how to config the "groupData" parameter in the context of jvm-packages. 
2: what is the "groupData" data format, just a text file or a rdd?

**Comment by superbobry** [comment]
Yes, in order to construct `Seq[Seq[Int]]` you need to know the partitioning of the RDD. This was the problem I was referring to. 

Do you think we can just add group column to the df and aggregate it later for a specific partition?

**Comment by anddelu** [comment]
Thanks all. According to docs, make sure that the partitions of groupData rdd correspond to partition of training data rdd before configing the parameter "groupData".

**Comment by anddelu** [comment]
I met a error "rank_obj.cc:50: Check failed: gptr.size() != 0 && gptr.back() == info.labels.size() group structure not consistent with #rows" when trained lambdarank model. It means that the paramter "groupData" is set mistakenly. I make "groupData" like this: for each partition rdd, I use the partitionId as index and use different group beginning position to make the format Seq[Int], lastly add the datas length to the Seq[Int]. For example: partition 0: data:["a", "a", "b", "b", "c"], I would get the Seq[Int] = Seq[0, 2, 4, 5].  can anybody tell me how to correctly generate the "groupData" format, otherwise why i met the problem?

**Comment by ebernhardson** [comment]
learning to rank works reasonably for me, although it is certainly fragile. For group generation I'm doing it like this: https://github.com/wikimedia/search-MjoLniR/blob/master/jvm/src/main/scala/org/wikimedia/search/mjolnir/PythonUtils.scala#L41

**Comment by anddelu** [comment]
thanks, @ebernhardson. I also trained model successfully.  before I misunderstand the groupdata,  now, for the partition 0: data:["a", "a", "b", "b", "c"], I should get the Seq[Int] = Seq[2, 2, 1]. It's really fragile when I load the saved data to train and meet error again. The training process will have a repartition action when numworkers isn't equal the partition nums. Before using your calcQueryGroups functions to make sure dataset not need to repartition so that the groupdata would make sense.

**Comment by mkolarek** [comment]
@ebernhardson I'm trying out your solution but I also get the "rank_obj.cc:50: Check failed: gptr.size() != 0 && gptr.back() == info.labels.size() group structure not consistent with #rows" error, is it maybe because of how I did the partitioning and sorting before? because I simply did ds.repartition($"queryIdCol").sort(queryIdCol). I also have more than 200 unique values in queryIdCol, would your method still be valid for this use case? sorry for reviving this issue
